File: 0755-pour-water/0755-pour-water.cpp
class Solution {
public:
    vector<int> pourWater(vector<int>& heights, int v, int k) {
        while(v--) drop(heights, k);
        return heights;
    }
    
    void drop(vector<int>& heights, int k) {
        int best = k;
        for(int d = -1; d <= 1; d += 2) {
            int i = k + d;
            while(i >= 0 && i < heights.size() && heights[i] <= heights[i-d]) {
                if(heights[i] < heights[best]) best = i;
                i += d;
            }
            if(best != k) break;
        }
        ++heights[best];
    }
};

File: 0755-pour-water/0755-pour-water.py
class Solution:
    def pourWater(self, heights: List[int], volume: int, k: int) -> List[int]:
        def drop(h, k):
            best = k
            for d in [-1,1]:
                i = k+d
                while i >= 0 and i < len(h) and h[i] <= h[i-d]:
                    if h[i] < h[best]: best = i
                    i += d
                if best != k: break
            h[best] += 1
        
        for _ in range(volume):
            drop(heights, k)
        return heights